core/src/main/java/com/linecorp/armeria/server/AbstractBindingBuilder.java

```diff
@@ -0,0 +1,316 @@
+
+        for (HttpMethod method : methods) {
+            if (!methodSet.add(method)) {
+                throw new IllegalArgumentException("duplicate HTTP method: " + method +
+                                                   ", for: " + routeBuilder);
+            }
+        }
+    }
+
+    /**
+     * Sets the {@link HttpMethod}s that a {@link Service} will support. If not set,
+     * {@link HttpMethod#knownMethods()}s are set.
+     *
+     * @see #path(String)
+     * @see #pathPrefix(String)
+     */
+    public AbstractBindingBuilder methods(HttpMethod... methods) {
+        return methods(ImmutableSet.copyOf(requireNonNull(methods, "methods")));
+    }
+
+    /**
+     * Sets the {@link HttpMethod}s that a {@link Service} will support. If not set,
+     * {@link HttpMethod#knownMethods()}s are set.
+     *
+     * @see #path(String)
+     * @see #pathPrefix(String)
+     */
+    public AbstractBindingBuilder methods(Iterable<HttpMethod> methods) {
+        requireNonNull(methods, "methods");
+        checkArgument(!Iterables.isEmpty(methods), "methods can't be empty");
+        this.methods = Sets.immutableEnumSet(methods);
+        return this;
+    }
+
+    /**
+     * Sets {@link MediaType}s that a {@link Service} will consume. If not set, the {@link Service}
+     * will accept all media types.
+     */
+    public AbstractBindingBuilder consumes(MediaType... consumeTypes) {
+        consumes(ImmutableSet.copyOf(requireNonNull(consumeTypes, "consumeTypes")));
+        return this;
+    }
+
+    /**
+     * Sets {@link MediaType}s that a {@link Service} will consume. If not set, the {@link Service}
+     * will accept all media types.
+     */
+    public AbstractBindingBuilder consumes(Iterable<MediaType> consumeTypes) {
+        ensureUniqueMediaTypes(consumeTypes, "consumeTypes");
+        this.consumeTypes = ImmutableSet.copyOf(consumeTypes);
+        return this;
+    }
+
+    /**
+     * Sets {@link MediaType}s that a {@link Service} will produce to be used in
+     * content negotiation. See <a href="https://tools.ietf.org/html/rfc7231#section-5.3.2">Accept header</a>
+     * for more information.
+     */
+    public AbstractBindingBuilder produces(MediaType... produceTypes) {
+        produces(ImmutableSet.copyOf(requireNonNull(produceTypes, "produceTypes")));
+        return this;
+    }
+
+    /**
+     * Sets {@link MediaType}s that a {@link Service} will produce to be used in
+     * content negotiation. See <a href="https://tools.ietf.org/html/rfc7231#section-5.3.2">Accept header</a>
+     * for more information.
+     */
+    public AbstractBindingBuilder produces(Iterable<MediaType> produceTypes) {
+        ensureUniqueMediaTypes(produceTypes, "produceTypes");
+        this.produceTypes = ImmutableSet.copyOf(produceTypes);
+        return this;
+    }
+
+    /**
+     * Returns a newly-created {@link Route}s based on the properties of this builder.
+     */
+    final List<Route> buildRouteList() {
```

Comment:
```pathBuilders.isEmpty() && !methods.isEmpty()``` is in `checkState` but... the validation logic is hard to read. 
It will be changed by https://github.com/line/armeria/pull/2040#discussion_r329858187